libraries/core_libs/consensus/src/vote_manager/vote_manager.cpp

```diff
@@ -471,6 +490,179 @@ uint64_t VoteManager::roundDeterminedFromVotes(size_t two_t_plus_one) {
   return 0;
 }
 
+void VoteManager::sendRewardPeriodCertVotes(uint64_t reward_period) {
+  auto reward_period_cert_votes = db_->getCertVotes(reward_period);
+  auto net = network_.lock();
+  net->onNewPbftVotes(std::move(reward_period_cert_votes), true);
+}
+
+bool VoteManager::AddRewardVote(const std::shared_ptr<Vote>& vote) {
+  // Don't check the vote if in DB here. That will check in updateRewardVotes for reducing DB read
+  const auto vote_hash = vote->getHash();
+  std::unique_lock lock(reward_votes_mutex_);
+  if (reward_votes_.second.contains(vote_hash)) {
+    return false;
+  }
+
+  const auto voted_block_hash = vote->getBlockHash();
+  const auto pbft_chain_last_block_hash = pbft_chain_->getLastPbftBlockHash();
+  if (voted_block_hash != pbft_chain_last_block_hash) {
+    LOG(log_dg_) << "Drop reward vote " << vote_hash << ". PBFT chain last block is " << pbft_chain_last_block_hash
+                 << ", but vote for PBFT block is " << voted_block_hash;
+    return false;
+  }
+
+  if (!reward_votes_.first) {
+    reward_votes_.first = voted_block_hash;
+  } else if (reward_votes_.first && reward_votes_.first != pbft_chain_last_block_hash) {
+    // Clear missing reward votes
+    LOG(log_dg_) << "Remove reward votes: \n";
+    for (const auto& v : reward_votes_.second) {
+      LOG(log_dg_) << v.first;
+    }
+    reward_votes_ = {};
+    reward_votes_.first = voted_block_hash;
+  }
+  assert(voted_block_hash == reward_votes_.first);
+
+  reward_votes_.second.insert({vote_hash, vote});
+  LOG(log_nf_) << "add reward vote " << vote_hash;
+
+  return true;
+}
+
+bool VoteManager::verifyRewardVote(std::shared_ptr<Vote>& vote) {
```

Comment:
why this is not happening inside `AddRewardVote` ? all the time we receive vote from network it should be verified

Reply:
if it was a part of "validated" votes in manager this does need to be run

Reply:
1. The main reason, PBFT chain is async with execution for final chain. And verify votes need DPOS states. We cannot get DPOS states if PBFT chain is much ahead of final chain. The situation is handled in PBFT manager. So we cannot verify votes in packet handler
2. The small reason, for performance, I prefer we do simple and important verification in packet handler, and do complecated verification in PBFT manger. Because PBFT manager is a timer machine, most of time is just at sleeping, not require performance. But packet handler do need performance

Reply:
I don't agree with you on 2
Even in transactions and DAG blocks design, we have 2 queues, one for unverified trx/DAG blocks, and one for verified trx/DAG blocks. In your option, we should remove unverified queues and do all verification in packet handlers.

Reply:
@gaoqi8649 it is nice that you dont agree but you are wrong. The whole purpose of our new priority queue for packets and processing threadpool is to put as much work as possible into the handlers. This is a fact - not something that you should agree or disagree

Reply:
put everything that can be into the packet handler